vm/src/vm.rs

```diff
@@ -289,23 +291,28 @@ impl VirtualMachine {
     {
         let args = args.into();
         trace!("Invoke: {:?} {:?}", func_ref, args);
-        match func_ref.payload {
-            PyObjectPayload::RustFunction { ref function } => function(self, args),
-            PyObjectPayload::Function {
-                ref code,
-                ref scope,
-                ref defaults,
-            } => self.invoke_python_function(code, scope, defaults, args),
-            PyObjectPayload::BoundMethod {
-                ref function,
-                ref object,
-            } => self.invoke(function.clone(), args.insert(object.clone())),
-            ref payload => {
-                // TODO: is it safe to just invoke __call__ otherwise?
-                trace!("invoke __call__ for: {:?}", payload);
-                self.call_method(&func_ref, "__call__", args)
-            }
+        if let Some(PyFunction {
```

Comment:
Why is a match statement no option here? I feel match is better than if statements with returns.

Reply:
I can get rid of the explicit returns, but a `match` won't work here because each `payload` call is generic over the specific payload type. 

(I do have some ideas on how to make this kind of thing cleaner and more efficient, but they're not fully baked yet.)

Reply:
I think we could use a macro, something like `match_payload!`, which would look like:

```
let match_res = match_payload!(pyobjref, {
    objint::PyInt(int) => do_thing(int.value),
    objfloat::PyFloat(objfloat::PyFloat { value }) => do_float_thing(value),
});
```